// db-access/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = void 0;
const logger_1 = require("@aws-lambda-powertools/logger");
const client_secrets_manager_1 = require("@aws-sdk/client-secrets-manager");
const postgresql_client_1 = require("postgresql-client");
const logging_util_1 = require("./logging-util");
const fs_1 = require("fs");
const PointCollectorLogFormatter_1 = require("./PointCollectorLogFormatter");
const resolveSecret = async (secretsManager, secretId) => {
    const secretValue = await secretsManager.getSecretValue({
        SecretId: secretId
    });
    if (secretValue.SecretString === undefined) {
        throw new Error(`No secret string for ${secretId}`);
    }
    const secret = JSON.parse(secretValue.SecretString);
    return secret;
};
const getConnection = (secret) => new postgresql_client_1.Connection({
    host: secret.host,
    port: secret.port,
    user: secret.username,
    password: secret.password,
    database: secret.dbname,
    ssl: {
        rejectUnauthorized: false,
        ca: (0, fs_1.readFileSync)('./eu-west-1-bundle.cert').toString() // cert from https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/UsingWithRDS.SSL.html
    }
});
const logger = new logger_1.Logger({ serviceName: (0, logging_util_1.serviceName)('db-setup'), logFormatter: new PointCollectorLogFormatter_1.PointCollectorLogFormatter() });
const handler = async (event, context) => {
    logger.info('EVENT: \n' + JSON.stringify(event, null, 2));
    logger.addContext(context);
    if (process.env.REGION === undefined ||
        process.env.DB_SCHEMA_NAME === undefined ||
        process.env.DB_USER_SECRET_ARN == undefined) {
        throw new Error('Missing environment variable (either REGION, DB_SECRET_ARN or DB_SCHEMA_NAME or DB_USER_SECRET_ARN');
    }
    try {
        const REGION = process.env.REGION;
        //const SCHEMA = process.env.DB_SCHEMA_NAME
        const secretsManager = new client_secrets_manager_1.SecretsManager({ region: REGION });
        const dbUserSecret = await resolveSecret(secretsManager, process.env.DB_USER_SECRET_ARN);
        logger.info('Opening connection...');
        const connection = getConnection(dbUserSecret);
        await connection.connect();
        const queries = event.queries;
        for (const query of queries) {
            if (query.startsWith('#')) {
                console.log('Skipping query: ' + query);
                continue;
            }
            console.log('Executing query: ' + query);
            const result = await connection.query(query);
            console.log(JSON.stringify(result, null, 2));
        }
        await connection.close();
    }
    catch (error) {
        logger.error('error', (0, logging_util_1.logData)(error));
        throw error;
    }
    return {
        PhysicalResourceId: 'DBAccess',
        Data: {
            Response: 'Successfully executed statements'
        }
    };
};
exports.handler = handler;
