// @aws-lambda-powertools/logger/lib/formatter/LogFormatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogFormatter = void 0;
const isErrorWithCause = (error) => {
    return 'cause' in error;
};
class LogFormatter {
    formatError(error) {
        return {
            name: error.name,
            location: this.getCodeLocation(error.stack),
            message: error.message,
            stack: error.stack,
            cause: isErrorWithCause(error)
                ? error.cause instanceof Error
                    ? this.formatError(error.cause)
                    : error.cause
                : undefined,
        };
    }
    formatTimestamp(now) {
        return now.toISOString();
    }
    getCodeLocation(stack) {
        if (!stack) {
            return '';
        }
        const stackLines = stack.split('\n');
        const regex = /\((.*):(\d+):(\d+)\)\\?$/;
        let i;
        for (i = 0; i < stackLines.length; i++) {
            const match = regex.exec(stackLines[i]);
            if (Array.isArray(match)) {
                return `${match[1]}:${Number(match[2])}`;
            }
        }
        return '';
    }
}
exports.LogFormatter = LogFormatter;
