// @aws-lambda-powertools/commons/lib/awsSdk/userAgentMiddleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addUserAgentMiddleware = exports.customUserAgentMiddleware = void 0;
const version_1 = require("../version");
const utils_1 = require("./utils");
const EXEC_ENV = process.env.AWS_EXECUTION_ENV || 'NA';
const middlewareOptions = {
    relation: 'after',
    toMiddleware: 'getUserAgentMiddleware',
    name: 'addPowertoolsToUserAgent',
    tags: ['POWERTOOLS', 'USER_AGENT'],
};
const customUserAgentMiddleware = (feature) => {
    return (next) => async (args) => {
        const powertoolsUserAgent = `PT/${feature}/${version_1.PT_VERSION} PTEnv/${EXEC_ENV}`;
        args.request.headers['user-agent'] = `${args.request.headers['user-agent']} ${powertoolsUserAgent}`;
        return await next(args);
    };
};
exports.customUserAgentMiddleware = customUserAgentMiddleware;
const addUserAgentMiddleware = (client, feature) => {
    try {
        if ((0, utils_1.isSdkClient)(client)) {
            if (client.middlewareStack
                .identify()
                .includes('addPowertoolsToUserAgent: POWERTOOLS,USER_AGENT')) {
                return;
            }
            client.middlewareStack.addRelativeTo(customUserAgentMiddleware(feature), middlewareOptions);
        }
        else {
            throw new Error(`The client provided does not match the expected interface`);
        }
    }
    catch (e) {
        console.warn('Failed to add user agent middleware', e);
    }
};
exports.addUserAgentMiddleware = addUserAgentMiddleware;
